OpenRA.Mods.Common/Traits/CaptureManager.cs

```diff
@@ -0,0 +1,278 @@
+					neutralCapturableTypes = neutralCapturableTypes.Union(c.Info.Types);
+
+				if (c.Info.ValidStances.HasStance(Stance.Enemy))
+					enemyCapturableTypes = enemyCapturableTypes.Union(c.Info.Types);
+			}
+		}
+
+		public void RefreshCaptures(Actor self)
+		{
+			capturesTypes = enabledCaptures.Aggregate(
+				default(BitSet<CaptureType>),
+				(a, b) => a.Union(b.Info.CaptureTypes));
+		}
+
+		public bool CanBeTargetedBy(Actor self, Actor captor, CaptureManager captorManager)
+		{
+			var stance = self.Owner.Stances[captor.Owner];
+			if (stance.HasStance(Stance.Enemy))
+				return captorManager.capturesTypes.Overlaps(enemyCapturableTypes);
+
+			if (stance.HasStance(Stance.Neutral))
+				return captorManager.capturesTypes.Overlaps(neutralCapturableTypes);
+
+			if (stance.HasStance(Stance.Ally))
+				return captorManager.capturesTypes.Overlaps(allyCapturableTypes);
+
+			return false;
+		}
+
+		public bool CanBeTargetedBy(Actor self, Actor captor, Captures captures)
+		{
+			if (captures.IsTraitDisabled)
+				return false;
+
+			var stance = self.Owner.Stances[captor.Owner];
+			if (stance.HasStance(Stance.Enemy))
+				return captures.Info.CaptureTypes.Overlaps(enemyCapturableTypes);
+
+			if (stance.HasStance(Stance.Neutral))
+				return captures.Info.CaptureTypes.Overlaps(neutralCapturableTypes);
+
+			if (stance.HasStance(Stance.Ally))
+				return captures.Info.CaptureTypes.Overlaps(allyCapturableTypes);
+
+			return false;
+		}
+
+		public Captures ValidCapturesWithLowestSabotageThreshold(Actor self, Actor captee, CaptureManager capteeManager)
+		{
+			if (captee.IsDead)
+				return null;
+
+			foreach (var c in enabledCaptures.OrderBy(c => c.Info.SabotageThreshold))
+				if (capteeManager.CanBeTargetedBy(captee, self, c))
+					return c;
+
+			return null;
+		}
+
+		public void OnCapture(Actor self, Actor captor, Player oldOwner, Player newOwner)
+		{
+			BeingCaptured = true;
+			self.World.AddFrameEndTask(w => BeingCaptured = false);
+		}
+
+		/// <summary>
+		/// Called by CaptureActor when the activity is ready to enter and capture the target.
+		/// This method grants the capturing conditions on the captor and target and returns
+		/// true if the captor is able to start entering or false if it needs to wait.
+		/// </summary>
+		public bool TryCapture(Actor self, Actor target, CaptureManager targetManager, out Captures captures)
+		{
+			captures = null;
+
+			if (target != currentTarget)
+			{
+				if (currentTarget != null)
+					CancelCapture(self, currentTarget, currentTargetManager);
+
+				targetManager.currentCaptors.Add(self);
+				currentTarget = target;
+				currentTargetManager = targetManager;
+				currentTargetDelay = 0;
+			}
+			else
+				currentTargetDelay += 1;
+
+			if (conditionManager != null && !string.IsNullOrEmpty(info.CapturingCondition) &&
+					capturingToken == ConditionManager.InvalidConditionToken)
+				capturingToken = conditionManager.GrantCondition(self, info.CapturingCondition);
```

Comment:
Why do we grant the condition even if `captures == null` below (and the method returns `false`)?

Reply:
From the method description
```csharp
/// Called by CaptureActor when the activity is ready to enter and capture the target.
/// This method grants the capturing conditions on the captor and target and returns
/// true if the captor is able to start entering or false if it needs to wait.
```
The conditions are active for the whole capture period (including the delay).  If we return false because the target is no longer capturable, then the code from `CaptureActor` will kick in to cancel the activity, and the condition will be disabled at the appropriate time.

Reply:
Renamed `Enter.CanEnter` to `Enter.TryStartEnter` and `CaptureManager.TryCapture` to `CaptureManager.StartCapture` to hopefully fix this confusion.